File: nosql-create-index-go/dataplane.go
package main

import (
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"os"
	"regexp"
	"sort"
	"strings"
	"sync"
	"time"

	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
	"github.com/Azure/azure-sdk-for-go/sdk/azcore/policy"
	"github.com/Azure/azure-sdk-for-go/sdk/data/azcosmos"
)

// Set to 1 for sequential insertion (helps with index stability)
const maxInsertConcurrency = 1

var validIdentifier = regexp.MustCompile(`^[A-Za-z_][A-Za-z0-9_]*$`)

type InsertStats struct {
	Total         int
	Inserted      int
	Skipped       int
	Failed        int
	RequestCharge float64
}

type VectorSearchResult struct {
	HotelID     string  `json:"HotelId"`
	HotelName   string  `json:"HotelName"`
	Description string  `json:"Description"`
	Score       float64 `json:"SimilarityScore"`
}

type ContainerRunResult struct {
	ContainerName string
	InsertStats   InsertStats
	Results       []VectorSearchResult
	RequestCharge float64
}

type embeddingsRequest struct {
	Input string `json:"input"`
}

type embeddingsResponse struct {
	Data []struct {
		Embedding []float64 `json:"embedding"`
	} `json:"data"`
	Error *struct {
		Message string `json:"message"`
	} `json:"error,omitempty"`
}

func LoadDocuments(path string) ([]map[string]any, error) {
	raw, err := os.ReadFile(path)
	if err != nil {
		return nil, fmt.Errorf("read data file: %w", err)
	}

	var documents []map[string]any
	if err := json.Unmarshal(raw, &documents); err != nil {
		return nil, fmt.Errorf("parse data file: %w", err)
	}

	// Validate Region property and prepare documents
	validRegions := map[string]bool{"Northeast": true, "Midwest": true, "South": true, "West": true}
	regionsFound := make(map[string]bool)

	prepared := make([]map[string]any, 0, len(documents))
	for idx, doc := range documents {
		hotelID, _ := doc["HotelId"].(string)
		if strings.TrimSpace(hotelID) == "" {
			return nil, errors.New("every document must include a non-empty HotelId")
		}

		// Validate Region property
		region, ok := doc["Region"].(string)
		if !ok || strings.TrimSpace(region) == "" {
			return nil, fmt.Errorf("document at index %d (HotelId=%s) missing or invalid Region property", idx, hotelID)
		}
		if !validRegions[region] {
			return nil, fmt.Errorf("document at index %d (HotelId=%s) has unexpected Region '%s'", idx, hotelID, region)
		}
		regionsFound[region] = true

		preparedDoc := cloneDocument(doc)
		preparedDoc["id"] = hotelID
		// Do NOT overwrite Region — keep it as-is for partition key
		prepared = append(prepared, preparedDoc)
	}

	// Log region distribution
	regions := make([]string, 0, len(regionsFound))
	for r := range regionsFound {
		regions = append(regions, r)
	}
	sort.Strings(regions)
	fmt.Printf("✓ Region validation passed. Found regions: %s\n", strings.Join(regions, ", "))

	// Count documents per region
	regionCounts := make(map[string]int)
	for _, doc := range prepared {
		region := doc["Region"].(string)
		regionCounts[region]++
	}
	// Print per-region counts in order
	for _, region := range []string{"Northeast", "Midwest", "South", "West"} {
		if count, exists := regionCounts[region]; exists {
			fmt.Printf("  Region '%s': %d documents\n", region, count)
		}
	}

	return prepared, nil
}

func cloneDocument(source map[string]any) map[string]any {
	clone := make(map[string]any, len(source)+2)
	for key, value := range source {
		clone[key] = value
	}
	return clone
}

func GenerateEmbedding(ctx context.Context, httpClient *http.Client, credential azcore.TokenCredential, cfg *Config, text string) ([]float32, error) {
	token, err := credential.GetToken(ctx, policy.TokenRequestOptions{Scopes: []string{"https://cognitiveservices.azure.com/.default"}})
	if err != nil {
		return nil, fmt.Errorf("get Azure OpenAI bearer token: %w", err)
	}

	requestBody, err := json.Marshal(embeddingsRequest{Input: text})
	if err != nil {
		return nil, fmt.Errorf("marshal embeddings request: %w", err)
	}

	endpoint, err := url.Parse(strings.TrimRight(cfg.OpenAIEmbeddingEndpoint, "/"))
	if err != nil {
		return nil, fmt.Errorf("parse AZURE_OPENAI_EMBEDDING_ENDPOINT: %w", err)
	}
	endpoint.Path = fmt.Sprintf("/openai/deployments/%s/embeddings", cfg.OpenAIEmbeddingDeployment)
	query := endpoint.Query()
	query.Set("api-version", cfg.OpenAIAPIVersion)
	endpoint.RawQuery = query.Encode()

	request, err := http.NewRequestWithContext(ctx, http.MethodPost, endpoint.String(), bytes.NewReader(requestBody))
	if err != nil {
		return nil, fmt.Errorf("create embeddings request: %w", err)
	}
	request.Header.Set("Authorization", "Bearer "+token.Token)
	request.Header.Set("Content-Type", "application/json")

	response, err := httpClient.Do(request)
	if err != nil {
		return nil, fmt.Errorf("call Azure OpenAI embeddings API: %w", err)
	}
	defer response.Body.Close()

	responseBody, err := io.ReadAll(response.Body)
	if err != nil {
		return nil, fmt.Errorf("read embeddings response: %w", err)
	}

	if response.StatusCode < http.StatusOK || response.StatusCode >= http.StatusMultipleChoices {
		return nil, fmt.Errorf("Azure OpenAI embeddings request failed: status=%d body=%s", response.StatusCode, strings.TrimSpace(string(responseBody)))
	}

	var payload embeddingsResponse
	if err := json.Unmarshal(responseBody, &payload); err != nil {
		return nil, fmt.Errorf("parse embeddings response: %w", err)
	}
	if payload.Error != nil {
		return nil, fmt.Errorf("Azure OpenAI embeddings error: %s", payload.Error.Message)
	}
	if len(payload.Data) == 0 {
		return nil, errors.New("Azure OpenAI embeddings response did not include any vectors")
	}

	embedding := make([]float32, len(payload.Data[0].Embedding))
	for index, value := range payload.Data[0].Embedding {
		embedding[index] = float32(value)
	}
	if len(embedding) != cfg.EmbeddingDimensions {
		return nil, fmt.Errorf("embedding dimensions mismatch: got %d, expected %d", len(embedding), cfg.EmbeddingDimensions)
	}

	return embedding, nil
}

func InsertDocuments(ctx context.Context, container *azcosmos.ContainerClient, documents []map[string]any) (InsertStats, error) {
	stats := InsertStats{Total: len(documents)}
	semaphore := make(chan struct{}, maxInsertConcurrency)
	var mu sync.Mutex
	var wg sync.WaitGroup

	for _, document := range documents {
		document := document
		wg.Add(1)
		go func() {
			defer wg.Done()
			semaphore <- struct{}{}
			defer func() { <-semaphore }()

			// Extract Region from document for partition key
			region, ok := document["Region"].(string)
			if !ok {
				mu.Lock()
				stats.Failed++
				mu.Unlock()
				return
			}
			partitionKey := azcosmos.NewPartitionKey().AppendString(region)

			body, err := json.Marshal(document)
			if err != nil {
				mu.Lock()
				stats.Failed++
				mu.Unlock()
				return
			}

			// Retry with backoff on 429 (rate limiting)
			maxRetries := 5
			for attempt := 0; attempt <= maxRetries; attempt++ {
				response, err := container.CreateItem(ctx, partitionKey, body, nil)
				if err != nil {
					var responseError *azcore.ResponseError
					if errors.As(err, &responseError) {
						if responseError.StatusCode == http.StatusConflict {
							mu.Lock()
							stats.Skipped++
							mu.Unlock()
							return
						}
						if responseError.StatusCode == http.StatusTooManyRequests && attempt < maxRetries {
							wait := time.Duration(attempt+1) * time.Second
							time.Sleep(wait)
							continue
						}
					}
					mu.Lock()
					stats.Failed++
					mu.Unlock()
					return
				}

				mu.Lock()
				stats.Inserted++
				stats.RequestCharge += float64(response.RequestCharge)
				mu.Unlock()
				return
			}
		}()
	}

	wg.Wait()
	if stats.Failed > 0 {
		return stats, fmt.Errorf("failed to insert %d documents", stats.Failed)
	}

	return stats, nil
}

// CountDocuments counts documents in a specific partition
func CountDocuments(ctx context.Context, container *azcosmos.ContainerClient, partitionKeyValue string) (int, error) {
	queryText := `SELECT VALUE COUNT(1) FROM c WHERE c.Region = @partitionKey`
	options := azcosmos.QueryOptions{
		QueryParameters: []azcosmos.QueryParameter{
			{Name: "@partitionKey", Value: partitionKeyValue},
		},
	}
	partitionKey := azcosmos.NewPartitionKey().AppendString(partitionKeyValue)
	pager := container.NewQueryItemsPager(queryText, partitionKey, &options)

	for pager.More() {
		page, err := pager.NextPage(ctx)
		if err != nil {
			return 0, fmt.Errorf("count query: %w", err)
		}
		for _, item := range page.Items {
			var count int
			if err := json.Unmarshal(item, &count); err != nil {
				return 0, fmt.Errorf("parse count: %w", err)
			}
			return count, nil
		}
	}
	return 0, nil
}

func QueryTopHotels(ctx context.Context, container *azcosmos.ContainerClient, embedding []float32, embeddingField, distanceFunction, partitionKeyValue string) ([]VectorSearchResult, float64, error) {
	if !validIdentifier.MatchString(embeddingField) {
		return nil, 0, fmt.Errorf("invalid embedding field name: %q", embeddingField)
	}

	// Convert float32 to float64 for parameter binding
	// The Cosmos DB SDK expects []float64 for numeric array parameters
	embedding64 := make([]float64, len(embedding))
	for i, v := range embedding {
		embedding64[i] = float64(v)
	}

	// ORDER BY VectorDistance(...) is required to rank results nearest-first.
	// Without it, SELECT TOP 5 returns 5 arbitrary documents from the partition,
	// not the nearest neighbors. SDK partition-key routing scopes the search to a
	// single partition, so the SQL stays focused on vector ranking.
	queryText := fmt.Sprintf(`SELECT TOP 5
		c.HotelId,
		c.HotelName,
		c.Description,
		VectorDistance(c.%s, @embedding, false, {'distanceFunction': '%s'}) AS SimilarityScore
	FROM c
	ORDER BY VectorDistance(c.%s, @embedding, false, {'distanceFunction': '%s'})`, embeddingField, distanceFunction, embeddingField, distanceFunction)

	options := azcosmos.QueryOptions{
		QueryParameters: []azcosmos.QueryParameter{
			{
				Name:  "@embedding",
				Value: embedding64,
			},
		},
	}

	// SDK-level partition key routing scopes the vector query to one partition.
	partitionKey := azcosmos.NewPartitionKey().AppendString(partitionKeyValue)
	pager := container.NewQueryItemsPager(queryText, partitionKey, &options)

	var results []VectorSearchResult
	var totalRequestCharge float64

	for pager.More() {
		page, err := pager.NextPage(ctx)
		if err != nil {
			return nil, totalRequestCharge, fmt.Errorf("query page: %w", err)
		}
		totalRequestCharge += float64(page.RequestCharge)
		for _, item := range page.Items {
			var result VectorSearchResult
			if err := json.Unmarshal(item, &result); err != nil {
				return nil, totalRequestCharge, fmt.Errorf("parse vector search result: %w", err)
			}
			results = append(results, result)
		}
	}

	return results, totalRequestCharge, nil
}

func RunContainerScenario(ctx context.Context, container *azcosmos.ContainerClient, containerName string, documents []map[string]any, embedding []float32, embeddingField, partitionKeyValue string) (*ContainerRunResult, error) {
	insertStats, err := InsertDocuments(ctx, container, documents)
	if err != nil {
		return nil, err
	}

	results, requestCharge, err := QueryTopHotels(ctx, container, embedding, embeddingField, "Cosine", partitionKeyValue)
	if err != nil {
		return nil, err
	}

	return &ContainerRunResult{
		ContainerName: containerName,
		InsertStats:   insertStats,
		Results:       results,
		RequestCharge: requestCharge,
	}, nil
}

func PrintContainerRunResult(result *ContainerRunResult) {
	fmt.Printf("\n=== %s ===\n", result.ContainerName)
	fmt.Printf("inserted=%d skipped=%d failed=%d total=%d writeRU=%.2f\n", result.InsertStats.Inserted, result.InsertStats.Skipped, result.InsertStats.Failed, result.InsertStats.Total, result.InsertStats.RequestCharge)
	for index, item := range result.Results {
		fmt.Printf("%d. HotelId=%s | HotelName=%s | score=%.4f | Description=%s\n", index+1, item.HotelID, item.HotelName, item.Score, summarizeDescription(item.Description))
	}
	fmt.Printf("queryRU=%.2f\n", result.RequestCharge)
}

func summarizeDescription(value string) string {
	trimmed := strings.Join(strings.Fields(value), " ")
	if len(trimmed) <= 88 {
		return trimmed
	}
	return trimmed[:85] + "..."
}

func algorithmLabel(containerName string) string {
	switch {
	case strings.Contains(containerName, "diskann"):
		return "DiskANN"
	case strings.Contains(containerName, "quantizedflat"):
		return "QuantizedFlat"
	case strings.Contains(containerName, "flat"):
		return "Flat"
	default:
		return containerName
	}
}

func dashes(n int) string {
	return strings.Repeat("-", n)
}

func truncate(s string, maxLen int) string {
	if len(s) <= maxLen {
		return s
	}
	return s[:maxLen-3] + "..."
}

func newHTTPClient() *http.Client {
	return &http.Client{Timeout: 60 * time.Second}
}

// ClearContainerData deletes all sample-inserted documents from a container
func ClearContainerData(ctx context.Context, container *azcosmos.ContainerClient) error {
	// Query across all regions to find all documents
	regions := []string{"Northeast", "Midwest", "South", "West"}

	for _, region := range regions {
		queryText := "SELECT c.id, c.Region FROM c WHERE c.Region = @region"
		options := azcosmos.QueryOptions{
			QueryParameters: []azcosmos.QueryParameter{
				{Name: "@region", Value: region},
			},
		}
		partitionKey := azcosmos.NewPartitionKey().AppendString(region)
		pager := container.NewQueryItemsPager(queryText, partitionKey, &options)

		for pager.More() {
			page, err := pager.NextPage(ctx)
			if err != nil {
				// Region may have no documents, continue
				break
			}
			for _, item := range page.Items {
				var doc map[string]any
				if err := json.Unmarshal(item, &doc); err != nil {
					return fmt.Errorf("parse item for deletion: %w", err)
				}
				if id, ok := doc["id"].(string); ok {
					_, err := container.DeleteItem(ctx, partitionKey, id, nil)
					if err != nil {
						return fmt.Errorf("delete item %q: %w", id, err)
					}
				}
			}
		}
	}

	return nil
}

File: nosql-create-index-java/src/main/java/com/azure/cosmos/createindex/DataPlane.java
package com.azure.cosmos.createindex;

import com.azure.ai.openai.OpenAIClient;
import com.azure.ai.openai.OpenAIClientBuilder;
import com.azure.ai.openai.models.EmbeddingsOptions;
import com.azure.core.credential.TokenCredential;
import com.azure.cosmos.CosmosClient;
import com.azure.cosmos.CosmosClientBuilder;
import com.azure.cosmos.CosmosContainer;
import com.azure.cosmos.models.CosmosBulkOperations;
import com.azure.cosmos.models.CosmosItemOperation;
import com.azure.cosmos.models.CosmosQueryRequestOptions;
import com.azure.cosmos.models.PartitionKeyBuilder;
import com.azure.cosmos.models.SqlParameter;
import com.azure.cosmos.models.SqlQuerySpec;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.io.IOException;
import java.nio.file.Files;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.regex.Pattern;

public final class DataPlane {
    private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();
    private static final Pattern FIELD_NAME_PATTERN = Pattern.compile("^[A-Za-z_][A-Za-z0-9_]*$");

    private DataPlane() {
    }

    public static CosmosClient createCosmosClient(TokenCredential credential, SampleConfig config) {
        return new CosmosClientBuilder()
                .endpoint(config.cosmosEndpoint())
                .credential(credential)
                .contentResponseOnWriteEnabled(false)
                .buildClient();
    }

    public static OpenAIClient createAzureOpenAIClient(TokenCredential credential, SampleConfig config) {
        return new OpenAIClientBuilder()
                .endpoint(config.openAiEmbeddingEndpoint())
                .credential(credential)
                .buildClient();
    }

    public static List<Float> generateEmbedding(OpenAIClient client, SampleConfig config, String text) {
        EmbeddingsOptions options = new EmbeddingsOptions(List.of(text));
        return client.getEmbeddings(config.openAiEmbeddingDeployment(), options)
                .getData()
                .get(0)
                .getEmbedding();
    }

    public static void verifyEmbeddingDimensions(OpenAIClient client, SampleConfig config) {
        List<Float> embedding = generateEmbedding(client, config, "dimension check");
        int actualDimensions = embedding.size();

        if (actualDimensions != config.expectedDimensions()) {
            throw new IllegalStateException(
                    "Embedding dimensions do not match the container definition. Expected "
                            + config.expectedDimensions() + ", received " + actualDimensions + ".");
        }
    }

    public static List<Map<String, Object>> readDocuments(SampleConfig config) throws IOException {
        byte[] payload = Files.readAllBytes(config.dataFileWithVectors());
        List<Map<String, Object>> items = OBJECT_MAPPER.readValue(
                payload,
                new TypeReference<List<Map<String, Object>>>() {
                });

        java.util.Set<String> regionsFound = new java.util.LinkedHashSet<>();
        java.util.Set<String> validRegions = java.util.Set.of("Northeast", "Midwest", "South", "West");

        List<Map<String, Object>> documents = new ArrayList<>(items.size());
        for (int idx = 0; idx < items.size(); idx++) {
            Map<String, Object> item = items.get(idx);
            LinkedHashMap<String, Object> document = new LinkedHashMap<>(item);

            // Validate HotelId
            Object hotelId = document.get("HotelId");
            if (hotelId == null) {
                throw new IllegalStateException("Each document must contain HotelId.");
            }

            // Validate Region property
            Object region = document.get("Region");
            if (region == null) {
                throw new IllegalStateException(
                        "Document at index " + idx + " (HotelId=" + hotelId + ") missing Region property");
            }
            String regionStr = String.valueOf(region);
            if (!validRegions.contains(regionStr)) {
                throw new IllegalStateException(
                        "Document at index " + idx + " (HotelId=" + hotelId + ") has unexpected Region '" + regionStr + "'");
            }
            regionsFound.add(regionStr);

            document.put("id", String.valueOf(hotelId));
            // Do NOT overwrite Region — keep it as-is for partition key
            documents.add(document);
        }

        // Log region distribution
        System.out.println("✓ Region validation passed. Found regions: " + String.join(", ", regionsFound));

        // Count documents per region
        Map<String, Integer> regionCounts = new HashMap<>();
        for (Map<String, Object> document : documents) {
            String region = (String) document.get("Region");
            regionCounts.put(region, regionCounts.getOrDefault(region, 0) + 1);
        }
        // Print per-region counts in order
        for (String region : Arrays.asList("Northeast", "Midwest", "South", "West")) {
            if (regionCounts.containsKey(region)) {
                System.out.println("  Region '" + region + "': " + regionCounts.get(region) + " documents");
            }
        }

        return documents;
    }

    public static IngestionSummary ingestDocuments(
            CosmosContainer container,
            String containerName,
            List<Map<String, Object>> documents) {

        // Group by region and print per-region counts as batch progress (matches .NET output)
        Map<String, List<Map<String, Object>>> byRegion = new java.util.TreeMap<>();
        for (Map<String, Object> doc : documents) {
            String region = String.valueOf(doc.get("Region"));
            byRegion.computeIfAbsent(region, k -> new ArrayList<>()).add(doc);
        }
        for (Map.Entry<String, List<Map<String, Object>>> entry : byRegion.entrySet()) {
            System.out.printf("  Region '%s': %d documents%n", entry.getKey(), entry.getValue().size());
        }

        List<CosmosItemOperation> operations = new ArrayList<>(documents.size());
        for (Map<String, Object> document : documents) {
            // Extract Region from document for partition key
            Object region = document.get("Region");
            if (region == null) {
                throw new IllegalStateException("Document missing Region property");
            }
            operations.add(CosmosBulkOperations.getUpsertItemOperation(
                    document,
                    new PartitionKeyBuilder().add(String.valueOf(region)).build()));
        }

        int upsertedDocuments = 0;
        int failedDocuments = 0;
        double requestCharge = 0.0;

        for (var response : container.executeBulkOperations(operations)) {
            var itemResponse = response.getResponse();
            if (itemResponse == null) {
                failedDocuments++;
                continue;
            }

            requestCharge += itemResponse.getRequestCharge();
            int statusCode = itemResponse.getStatusCode();
            if (statusCode >= 200 && statusCode < 300) {
                upsertedDocuments++;
            } else {
                failedDocuments++;
            }
        }

        return new IngestionSummary(containerName, documents.size(), upsertedDocuments, failedDocuments, requestCharge);
    }

    public static QuerySummary queryTopMatches(
            CosmosContainer container,
            String containerName,
            SampleConfig config,
            List<Float> queryEmbedding,
            String distanceFunction) {
        String embeddingField = validateFieldName(config.embeddingFieldName());
        // Scope the vector query to one partition through SDK request options so the SQL focuses on ranking.
        String queryText = "SELECT TOP @topK c.HotelId, c.HotelName, c.Description, "
                + "VectorDistance(c." + embeddingField + ", @embedding, false, {'distanceFunction': '" + distanceFunction + "'}) AS SimilarityScore "
                + "FROM c "
                + "ORDER BY VectorDistance(c." + embeddingField + ", @embedding, false, {'distanceFunction': '" + distanceFunction + "'})";

        List<QueryResult> results = new ArrayList<>();
        double requestCharge = 0.0;

        // Query single partition (config.partitionKeyValue) for efficiency
        String partitionKeyValue = config.partitionKeyValue();
        SqlQuerySpec querySpec = new SqlQuerySpec(
                queryText,
                List.of(
                        new SqlParameter("@topK", config.topCount()),
                        new SqlParameter("@embedding", toDoubleList(queryEmbedding))));

        CosmosQueryRequestOptions options = new CosmosQueryRequestOptions();
        options.setPartitionKey(new PartitionKeyBuilder().add(partitionKeyValue).build());

        for (var page : container.queryItems(querySpec, options, Map.class).iterableByPage()) {
            requestCharge += page.getRequestCharge();
            for (Object item : page.getResults()) {
                @SuppressWarnings("unchecked")
                Map<String, Object> result = (Map<String, Object>) item;
                results.add(new QueryResult(
                        String.valueOf(result.get("HotelId")),
                        String.valueOf(result.get("HotelName")),
                        String.valueOf(result.get("Description")),
                        ((Number) result.get("SimilarityScore")).doubleValue()));
            }
        }

        return new QuerySummary(containerName, requestCharge, results);
    }

    public static void printQuerySummary(QuerySummary summary, SampleConfig config) {
        System.out.println("\n=== Query results: " + summary.containerName() + " ("
                + Config.algorithmLabel(summary.containerName(), config) + ") ===");
        System.out.printf("Request charge: %.2f RUs%n", summary.requestCharge());

        int rank = 1;
        for (QueryResult result : summary.results()) {
            System.out.printf("%d. HotelId=%s | HotelName=%s | score=%.4f | Description=%s%n",
                    rank++,
                    result.hotelId(),
                    result.hotelName(),
                    result.score(),
                    shorten(result.description()));
        }
    }

    private static String validateFieldName(String fieldName) {
        if (!FIELD_NAME_PATTERN.matcher(fieldName).matches()) {
            throw new IllegalArgumentException("Invalid embedding field name: " + fieldName);
        }
        return fieldName;
    }

    private static List<Double> toDoubleList(List<Float> embedding) {
        List<Double> values = new ArrayList<>(embedding.size());
        for (Float value : embedding) {
            values.add(value.doubleValue());
        }
        return values;
    }

    private static String shorten(String value) {
        if (value == null || value.length() <= 110) {
            return value;
        }
        return value.substring(0, 107).trim() + "...";
    }

    public static void clearContainerData(CosmosContainer container) {
        try {
            SqlQuerySpec querySpec = new SqlQuerySpec("SELECT c.id FROM c");
            container.queryItems(querySpec, new CosmosQueryRequestOptions(), Map.class)
                    .stream()
                    .forEach(item -> {
                        String id = String.valueOf(item.get("id"));
                        container.deleteItem(id, new PartitionKeyBuilder()
                                .add("hotels")
                                .build(),
                                null);
                    });
        } catch (Exception e) {
            throw new RuntimeException("Failed to clear container data: " + e.getMessage(), e);
        }
    }
}

record IngestionSummary(
        String containerName,
        int totalDocuments,
        int upsertedDocuments,
        int failedDocuments,
        double requestCharge) {
}

record QueryResult(
        String hotelId,
        String hotelName,
        String description,
        double score) {
}

record QuerySummary(
        String containerName,
        double requestCharge,
        List<QueryResult> results) {
}
